File: 2_AddTwoNumbers.java
class AddTwoNumbers {
  public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
    return recursive(l1, l2, 0);
  }

  ListNode recursive(ListNode p, ListNode q, int res) {
    if (p==null && q==null && res==0) return null;
    res += (p==null?0:p.val) + (q==null?0:q.val);
    ListNode node = new ListNode(res % 10);
    node.next = recursive(p==null?p:p.next, q==null?q:q.next, res/10);
    return node;
  }
}

File: 2_AddTwoNumbers.cpp
#define _LL_DELETE_RETURN_POINTER
#include <local_leetcode.hpp>

class Solution {
  public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
      ListNode dummy;

      ListNode* p = &dummy;
      int acc = 0;
      while (l1 || l2 || acc) {
        int n1 = l1 ? l1->val : 0;
        int n2 = l2 ? l2->val : 0;
        int sum = n1 + n2 + acc;
        acc = sum / 10;
        p->next = new ListNode(sum%10);
        p = p->next;

        if (l1)
          l1 = l1->next;
        if (l2)
          l2 = l2->next;
      }

      return dummy.next;
    }
};


int main() {
  EXECS(Solution::addTwoNumbers);
  return 0;
}